### ISSUE TITLE
get_menu passes arguments to RoutableViewMixin.allow()

### ISSUE THREAD

**Issue by melvyn-sopacua** [yourlabs/crudlfap#6]
In short:
```

File "/home/melvyn/hg/crudlfap/src/crudlfap/routers.py" in get_menu
  177.             for v in self.views

File "/home/melvyn/hg/crudlfap/src/crudlfap/routers.py" in <listcomp>
  180.                 and v.factory(**kwargs)().allow(user)

Exception Type: TypeError at /
Exception Value: allow() takes 1 positional argument but 2 were given
```
v.factory(**kwargs)() = RoutableViewMixin and it's allow() doesn't accept arguments.

Triggered by the homepage:

```python
url(r'^$', generic.TemplateView.as_view(
        template_name='crudlfap/home.html'
    )),
```
And the list view generated by the router:

```
urlpatterns = crudlfap.Router(
    models.Alias, fields='__all__', menus=[_('aliases')],
    allow=lambda view, user: True,
).urlpatterns()
```

**Comment by melvyn-sopacua** [comment]
`allow()` is called before `as_view()` and before `dispatch()` so we have no self.request and we need to have the user passed in. Fix inc.

**Comment by jpic** [comment]
self.request is set when constructing the view object, it should have a user attribute. Does it work if you remove the user argument from lambda and then use view.request.user ?

**Comment by melvyn-sopacua** [comment]
That's not the problem. It's passed to RoutableViewMixin.allow() which doesn't accept an argument. It's weird though, there's a test for it that should basically do what get_menu() does and succeeds, so working on a test now that makes it fail.

**Comment by melvyn-sopacua** [comment]
Got it. The `ExampleView` overrides allow(), which makes it pass.

**Comment by jpic** [comment]
I'm sorry, this is all because I rushed that refactor on sunday night, I have just pushed the necessary fixes in master.

I'm open to debating about what we should have in 1.0: explicit user arg, or rely on the request object.

In our private project where we have the dirty poc for crudlfa+ we removed the user explicit arg and relied on request and it works and gives a feeling of "better". But honnestly I don't know what is best.

**Comment by jpic** [comment]
Should we continue considering that we're going to call allow() on a view which is provisioned with a request or revert the recent changes I've been rushing for this ?

**Comment by melvyn-sopacua** [comment]
I'm not sure:

- create requires global create permission on model
- update requires object level change permission (think: django-guardian or self_or_editor semantics)
- delete as update

Maybe a more flexible signature is needed where the base class just checks the user.is_staff but any number of arguments can be passed in to do permission checks.

**Comment by jpic** [comment]
Another case is when a view performs an action with more than one object. In this case, it's really up to the user to create the view with the attributes it'll need to execute allow() successfully, either during `dispatch()`, either during permission check from another view.

I wonder, do we really need arguments, or should we consider that it's up to the user to `factory()` or `__init__()` a view with the context necessary for `allow()` to run. If they don't, then `allow()` will crash.

**Comment by melvyn-sopacua** [comment]
We could do something similar as `get_context_data` as in `get_permission_context` and allow() requests that and gets the 'user' keyword out of it by default.

So:

```python

class RoutableViewMixin(object):
    def get_permission_context(self, **kwargs):
        kwargs.setdefault('user', self.request.user)
        return kwargs

    def allow():
        ctx = self.get_permission_context()
        return ctx['user'].is_staff

class MyView(RoutableViewMixin, View):
    def get_permission_context(self, **kwargs):
        ctx = super().get_permission_context(**kwargs)
        ctx.update({'instance': self.get_object()})
        return ctx

**Comment by jpic** [comment]
Absolutely agreed, but I'm a bit lost, what are we trying to solve exactly ?

**Comment by melvyn-sopacua** [comment]
Have an allow method that doesn't require arguments but can still handle flexible permission checks.
The above even allows "edits only between 9 and 5" if you want to implement it.

**Comment by jpic** [comment]
Ok, but then, can you just pass the things that allow() needs to factory() or even `__init__()` ? Then it will be in self ?

**Comment by melvyn-sopacua** [comment]
Right, because we now define a view.

The hard case is the autogenerated views by the Router with only a model class reference.

**Comment by jpic** [comment]
That's exactly what the initial idea was like. 

Then it turned out the user was not sufficient, and we wanted to check stuff in the session too in the case when an admin user is impersonating another one or to check what proxy it's connecting with. So that's we we wanted to pass the request instead of the user. Then, why juggle to pass it to allow()  both from dispatch() and from outside when we could just pass the request to `__init__()` or `factory()`: we thought just always pass it at least through `__init__()` exactly like Django would.

Other factors made us think that:

- we made self.object a property that would call get_object() automatically, that works when allow() is called by dispatch(),
- we pass object=object in factory or `__init__()` so that `allow()` also works when called from elsewhere than dispatch (ie. templates),
- if django has this pattern where `__init__()` uses kwargs to set attributes, why not completely rely on this ?

I think the pro of this pattern is that it code golfs, the con is that it's confusing even for me haha

**Comment by melvyn-sopacua** [comment]
You're probably saying it needs real docs(tm). :D

**Comment by jpic** [comment]
But there's still something that's unfortunate. Because reverse() is a class method, it can't use things such as self.object even though it's called from a view instance that's been hydrated with the object variable, so we have to pass the object again to reverse(). Perhaps you would have an idea about this ?

**Comment by melvyn-sopacua** [comment]
I think you asked the right question before:

> what are we trying to solve exactly?

- **Router**: job is to create view classes and url patterns
- **View class**: job is have a class method `as_view()` that creates returns a callable
- **view callable**: creates an instance of the view class and calls it's `dispatch` method
- **url patterns**: a list of patterns that call the view callable
- **permissions**: must have access to request.user and validate a property or method on it

No where does `self.object` get in the way, except when you "abuse" the method flow from a template to hide/show a menu.
I think integrating [django-guardian](https://github.com/django-guardian/django-guardian/blob/devel/guardian/templatetags/guardian_tags.py) is a good idea. Especially since it also has a lightweight [permission mixin](https://github.com/django-guardian/django-guardian/blob/devel/guardian/mixins.py).

**Comment by jpic** [comment]
tl;dr I don't mind django-guardian support as long as it doesn't become a requirement: have never needed it, probably never will.

Should we add signals in models to update django-guardian permissions for every model then tell each view what django-guardian permission it requires and then check django-guardian permissions in each templates before every link display, and risk to have inconsistent permissions in the database table, instead of just implementing my permission logic with 1 SLOC OOAO in allow() + Router.get_queryset() ?

For me, the latter is good enough and feels safer and easier to maintain, that said, I have nothing against optional support for django-guardian, perhaps the easiest way would be to make a RouterGuardian which has an allow() and get_queryset() implementation which supports django-guardian ?

The purpose of allow() is to tell if a request is going to have permission to execute a view with a given context: from within dispatch as well as from within external views to help with UI. It was designed for that purpose and we have pretty complex permissions checks (with several privilege levels, different proxies, user impersonation ...) in production with that system so I don't believe guardian is going to be necessary for me any time soon.

But if we really want to make a solid opinion, perhaps we should compare both ?

- one app with django-guardian, 
- another with just Router.get_queryset() and allow() overrides.

Then you'll have all you need to decide what you want to have in production ;)

Currently, if you want to check that a user can execute the detail view of an object:

    # YourModel could also be 'yourapp.yourmodel' here
    view_class = Router.registry[YourModel]['detail']
    view = view_class(request=request, object=your_object)
    allowed = view.allow()

What about combining two objects ? 

    view_class = Router.registry[YourModel]['combine']
    view = view_class(request=request, object=your_object, yourcustom=other_object)
    allowed = view.allow()

Of course for just a list view, no need for any context:

    view_class = Router.registry[YourModel]['list']
    allowed = view_class(request=request).allow()

But if you don't like it we could have something in src/crudlfap_guardian, that would not hurt anyone ;)

Lol yeah Pic also means penis in some language don't remember which ^^

**Comment by jpic** [comment]
We use another system them, which supports django permissions by default. It's not complete, but this issue should not reoccur.

The router now has [django-guardian/django-rules/django compatible code](https://github.com/yourlabs/crudlfap/blob/master/src/crudlfap/router.py#L224) in a configurable way (set required_permissions or implement get_required_permissions thanks to [__getattr__ aggressivity]()).

The view may also override the allowed() method to decorate or override the default behaviour (calling the router allowed()).
